File: library/src/main/java/com/github/kongpf8848/rxhttp/request/UploadRequest.java
package com.github.kongpf8848.rxhttp.request;


import com.github.kongpf8848.rxhttp.HttpConstants;

import java.io.File;
import java.util.Iterator;
import java.util.Map;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

public class UploadRequest extends AbsRequest {

    public UploadRequest(String url) {
        super(url);
    }

    @Override
    public RequestBody buildRequestBody() {
        MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);
        Iterator<Map.Entry<String, Object>> iterator = params.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<String, Object> entry = iterator.next();
            String key = entry.getKey();
            Object value = entry.getValue();
            if (value instanceof File) {
                File file = (File) value;
                builder.addFormDataPart(key, file.getName(), RequestBody.create(MediaType.parse(HttpConstants.MIME_TYPE_BINARY), file));
            } else {
                builder.addFormDataPart(key, String.valueOf(value));
            }
        }
        return builder.build();

    }
}

File: RxHttp/src/main/java/io/github/kongpf8848/rxhttp/request/UploadRequest.kt
package io.github.kongpf8848.rxhttp.request

import android.app.Activity
import android.content.Context
import android.net.Uri
import androidx.fragment.app.Fragment
import io.github.kongpf8848.rxhttp.ByteArrayRequestBody
import io.github.kongpf8848.rxhttp.HttpConstants
import io.github.kongpf8848.rxhttp.Platform
import io.github.kongpf8848.rxhttp.ProgressRequestBody
import io.github.kongpf8848.rxhttp.callback.ProgressCallback
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File
import java.io.IOException

class UploadRequest : PostRequest {

    constructor(context: Context) : super(context)
    constructor(activity: Activity) : super(activity)
    constructor(fragment: Fragment) : super(fragment)

    override fun buildRequestBody(): RequestBody? {
        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        val params: Map<String, Any?>? = getParams()
        if (!params.isNullOrEmpty()) {
            for((key,value) in params){
                if (value is File) {
                    builder.addFormDataPart(key, value.name, RequestBody.create(HttpConstants.MIME_TYPE_BINARY.toMediaTypeOrNull(), value))
                } else if (value is Uri) {
                    try {
                        val inputStream = actualContext!!.contentResolver.openInputStream(value)
                        if (inputStream != null) {
                            val bytes = inputStream.readBytes()
                            inputStream.close()
                            builder.addFormDataPart(key, key, ByteArrayRequestBody(HttpConstants.MIME_TYPE_BINARY.toMediaTypeOrNull(), bytes))
                        }
                    } catch (e: IOException) {
                        e.printStackTrace()
                    }
                } else {
                    builder.addFormDataPart(key, value.toString())
                }
            }
        }
        return ProgressRequestBody(builder.build(), object : ProgressCallback {
            override fun onProgress(readBytes: Long, totalBytes: Long) {
                Platform.get().defaultCallbackExecutor()?.execute { callback?.onProgress(readBytes, totalBytes) }
            }
        })
    }
}